primitives/trie/src/recorder.rs

```diff
@@ -0,0 +1,377 @@
+		}
+	}
+
+	/// Convert the recording into a [`StorageProof`].
+	///
+	/// It requires the `root` and the `hash_db` to lookup values that we served from the `cache`
+	/// and hadn't recorded the trie nodes. It will lookup these values and ensure to record the
+	/// trie nodes to include them in the final [`StorageProof`].
+	///
+	/// While a recorder can be cloned, all share the same internal state. After calling this
+	/// function, all other instances will have their internal state reset as well.
+	///
+	/// Returns the [`StorageProof`] or an error if one of lookups in the trie failed.
+	pub fn into_storage_proof(
+		self,
+		root: &H::Out,
+		hash_db: &dyn HashDBRef<H, DBValue>,
+		cache: Option<&mut TrieCache<H>>,
+	) -> Result<StorageProof, crate::Error<H::Out>> {
+		let mut recorder = mem::take(&mut *self.inner.lock());
+		let accessed_keys = mem::take(&mut recorder.accessed_keys);
+		Self::record_accessed_keys(&mut recorder, accessed_keys, root, hash_db, cache)?;
+
+		Ok(StorageProof::new(recorder.accessed_nodes.drain().map(|(_, v)| v)))
+	}
+
+	/// Convert the recording to a [`StorageProof`].
+	///
+	/// It requires the `root` and the `hash_db` to lookup values that we served from the `cache`
+	/// and hadn't recorded the trie nodes. It will lookup these values and ensure to record the
+	/// trie nodes to include them in the final [`StorageProof`].
+	///
+	/// In contrast to [`Self::into_storage_proof`] this doesn't consumes and clears the recordings.
+	///
+	/// Returns the [`StorageProof`] or an error if one of lookups in the trie failed.
+	pub fn to_storage_proof(
+		&self,
+		root: &H::Out,
+		hash_db: &dyn HashDBRef<H, DBValue>,
+		cache: Option<&mut TrieCache<H>>,
+	) -> Result<StorageProof, crate::Error<H::Out>> {
+		let mut recorder = self.inner.lock();
+		let accessed_keys = recorder.accessed_keys.clone();
+		Self::record_accessed_keys(&mut *recorder, accessed_keys, root, hash_db, cache)?;
+		recorder
+			.accessed_keys
+			.values_mut()
+			.flat_map(|v| v.values_mut())
+			.for_each(|k| {
+				// Mark all trie nodes as recorded
+				k.trie_nodes_recorded = true;
+			});
+
+		Ok(StorageProof::new(recorder.accessed_nodes.iter().map(|(_, v)| v.clone())))
+	}
+
+	/// Record the trie nodes for the accessed keys.
+	fn record_accessed_keys(
+		inner: &mut RecorderInner<H::Out>,
+		accessed_keys: HashMap<H::Out, HashMap<Vec<u8>, AccessedKey>>,
+		root: &H::Out,
+		hash_db: &dyn HashDBRef<H, DBValue>,
+		mut cache: Option<&mut TrieCache<H>>,
+	) -> Result<(), crate::Error<H::Out>> {
+		tracing::trace!(
+			target: LOG_TARGET,
+			storage_root = ?root,
+			"Recording accessed keys"
+		);
+
+		let mut trie_recorder = TrieRecorder::<H, _> {
+			inner,
+			storage_root: *root,
+			encoded_size_estimation: Default::default(),
+		};
+
+		accessed_keys.into_iter().try_for_each(|(root, keys)| {
+			// For reading the trie the layout is compatible
+			let trie = TrieDBBuilder::<crate::LayoutV1<H>>::new(hash_db, &root)
```

Comment:
Yeah that is a really good comment. I added the functionality to also record the child tries, because before it didn't worked. 

Do we have any tests for `KeySpaceDB` that don't involve having a full rocksdb instance? Something in memory?